### PR TITLE
Add bayonets

### PR DESCRIPTION
**Pull request by notquitehadouken**
## About the PR

make it to where you can attach knives to (certain) guns
examine text on fixable guns shows if there is a bayonet attached, examining unfixable ones shows nothing

can fix onto anything that is a rifle or smg
can fix combat knife, survival knife, and e-dagger

## Why / Balance

Is cool.
Makes it easier to use melee with your gun since taking out a knife during close quarters combat is deceptively hard.
You cannot:
- Fix anything that isn't a combat knife, survival knife, or epen.
- Fix onto shotguns.
- Fix onto grenade launchers.
- Fix onto pistols.
- Fix onto machine guns.

The swing time will always be the same (1 second).

## Technical details

Allows wide swings on guns as long as the gun shoot with left click (see: crusher)
Adds the `MeleeRequiresObjectInSlot` component, which, when present, will only allow a melee weapon to be used if there is any item present in the slot specified.
Adds 2 variables to `MeleeWeaponComponent`, one for removing heavy attack capability and one for removing light attack capability.
The melee damage of the gun (the gun swings) when a bayonet is inserted is changed to the damage of the bayonet. This means admeme 9001 damage epen actually works as a bayonet
Adds the Bayonet tag, which allows a melee weapon to go into the bayonet slot in a gun.

## Media

https://github.com/space-wizards/space-station-14/assets/45953835/4deb4516-fefc-4ad4-98a2-60bac472b2ed

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- add: You can now attach combat knives, survival knives, and e-pens to the C-20r, Lecter, Drozd, AKMS, and WT550.

### CODE REVIEW

**Comment by UbaserB** [comment]
you forgot kitchen knives (they are technically knives)

**Comment by Tunguso4ka** [comment]
we need bayonet for the spear.

**Comment by UbaserB** [comment]
> we need bayonet for the spear.

isnt a spear WAY too big (except for like big big guns ig)

**Comment by Tunguso4ka** [comment]
> isnt a spear WAY too big (except for like big big guns ig)

i mean make knives attachable to the spears.
Why? Because it's fun.

**Comment by 27alaing** [comment]
what about the makeshift shivs

**Comment by UbaserB** [comment]
> > isnt a spear WAY too big (except for like big big guns ig)
> 
> i mean make knives attachable to the spears. Why? Because it's fun.

oh

**Comment by deltanedas** [comment]
> i mean make knives attachable to the spears.

human head spear bayonet :trollface:

**Comment by JoeHammad1844** [comment]
so bayonets will fit on the C20R, WT550 and PKA but not the lecter or L6 SAW?

**Comment by Endecc** [comment]
> so bayonets will fit on the C20R, WT550 and PKA but not the lecter or L6 SAW?

I think Lecter would make some sense to have one, also can't forget the customary Mosin bayonet!

**Comment by JoeHammad1844** [comment]
> you forgot kitchen knives (they are technically knives)

they probably wouldn't have a bayonet lug, which allows them to be attached

**Comment by deltanedas** [comment]
use duct tape on kitchen knife to attack it to a gun, but it has a chance of falling off when used

**Comment by KingOnBudget** [comment]
will we be able to attach e-daggers?

**Comment by JoeHammad1844** [comment]
how modular of a system is this? by that i mean how easy would it be to apply the same logic but with flashlights?

File: Content.Shared/Weapons/Melee/MeleeWeaponComponent.cs [hunk 1]
```diff
@@ -21,6 +21,18 @@ public sealed partial class MeleeWeaponComponent : Component
     [DataField, ViewVariables(VVAccess.ReadWrite)]
     public bool AltDisarm = true;
 
+    /// <summary>
+    /// If you can't light attack
+    /// </summary>
+    [DataField, ViewVariables(VVAccess.ReadOnly)]
```
Comment (deltanedas) [inline Content.Shared/Weapons/Melee/MeleeWeaponComponent.cs:27]: datafields are automatically vv ro, make it readwrite or remove the viewvariable attribute

**Comment by CatMagic1** [comment]
So... yeah. Is this a gun attachments system?

**Comment by notquitehadouken** [comment]
> so bayonets will fit on the C20R, WT550 and PKA but not the lecter or L6 SAW?

lecter yes, l6 no

File: Content.Shared/Weapons/Melee/MeleeWeaponComponent.cs (hunk 1 above)
Comment (notquitehadouken) [reply Content.Shared/Weapons/Melee/MeleeWeaponComponent.cs:27]: don't enough know why it's RO, will do

**Comment by notquitehadouken** [comment]
> So... yeah. Is this a gun attachments system?

it's just ItemSlotsSystem

> how modular of a system is this? by that i mean how easy would it be to apply the same logic but with flashlights?

maybe not as modular as i'd like it to be, but `MeleeRequiresObjectInSlot` is at least somewhat
depends on how easy it is to make lights come off of things with flashlights inside

**Comment by RiceMar1244** [comment]
To be honest I'm not a fan of the implementation here, its too hard coded and not modular in a way that would aid future additions of attachments

**Comment by notquitehadouken** [comment]
> To be honest I'm not a fan of the implementation here, its too hard coded and not modular in a way that would aid future additions of attachments
> 
> Unfortunate to say we might need some kind of attachment system before this can be implemented ideally :o

guy who was working on attachment system stopped working on it

**Comment by RiceMar1244** [comment]
> > To be honest I'm not a fan of the implementation here, its too hard coded and not modular in a way that would aid future additions of attachments
> > Unfortunate to say we might need some kind of attachment system before this can be implemented ideally :o
> 
> guy who was working on attachment system stopped working on it

To my knowledge, and I know what flareguy said, nobody was ever *actually* working on an attachment system. There was only conversation around maybe working on it. Nothing was ever shown for that or further discussed. Regardless, it still holds that in my own opinion, implementing bayonets as standalone feels a bit jank conceptually. Of course, I am not a maintainer, so this is only my input.

**Comment by notquitehadouken** [comment]
i'm Done™️, i just need to add media

**Comment by JoeHammad1844** [comment]
maybe leave the atreides, M90GL and vector out of the change log because they are currently unobtainable

**Comment by notquitehadouken** [comment]
> maybe leave the atreides, M90GL and vector out of the change log because they are currently unobtainable

yea

**Comment by notquitehadouken** [comment]
media added

**Comment by notquitehadouken** [comment]
not done

**Comment by JoeHammad1844** [comment]
maybe guns can do a melee attack without bayonets, but have it do like 6 damage

**Comment by notquitehadouken** [comment]
now i am done

**Comment by notquitehadouken** [comment]
now i am done

**Comment by RiceMar1244** [comment]
I've gotten the chance to more deeply peer-review the changes. I am not a maintainer, but I wanted to share some feedback because I believe there are some elements of the implementation that may be problematic and need consideration:
- The ItemSlotsSystem subscribes to an AttemptMeleeEvent. The problem is that this doesn't feel like the place for that, since its purpose is not related to the logic for item slots.
- The SharedGunSystem checks for an item slot of a hardcoded name. This results in an unintuitive interface, because devs will need to know that the item slot requires a specific name if they want their bayonets to work.
- Inserting a bayonet seems to copy the weapon's stats onto the gun's required MeleeWeaponComponent instead of making the knife itself usable from the gun. This is the biggest reason I believe in the necessity of a dedicated attachment system that would directly use any attached gear like flashlights or bayonets or under-barrel launches, instead of copying the functionality.

**Comment by notquitehadouken** [comment]
> I've gotten the chance to more deeply peer-review the changes. I am not a maintainer, but I wanted to share some feedback because I believe there are some elements of the implementation that may be problematic and need consideration:
> * The ItemSlotsSystem subscribes to an AttemptMeleeEvent. The problem is that this doesn't feel like the place for that, since its purpose is not related to the logic for item slots.

not sure why that's there, will move

> * The SharedGunSystem checks for an item slot of a hardcoded name. This results in an unintuitive interface, because devs will need to know that the item slot requires a specific name if they want their bayonets to work.

that's a failsafe. don't know why i did that.

> * Inserting a bayonet seems to copy the weapon's stats onto the gun's required MeleeWeaponComponent instead of making the knife itself usable from the gun. This is the biggest reason I believe in the necessity of a dedicated attachment system that would directly use any attached gear like flashlights or bayonets or under-barrel launches, instead of copying the functionality.

- the sprite that shows up when you swing the bayonet is that of the gun which is intended. making the gun swing instead of the bayonet is the most direct way to make this happen
- the swing speed is always 1 attack per second, and i'd rather only modify the gun instead of the bayonet. it makes keeping track of what has happened easier, and means the bayonet doesn't need to be changed again when it's removed.
- this pr is about bayonets. there are a lot of different things that have to be added in order for that to work, and if this was bundled into an all-in-one attachment system it would get really convoluted almost immediately. maybe attachments for flashlights, not for bayonets.

**Comment by notquitehadouken** [comment]
> There is no visual information for what a target is being attacked by. If these guns are really going to have no bayonet sprites on them, that means the only source of visual telling would have to be the melee attack's sprite, and youve intended it to still be the gun.

the difference is that you can see they are attacking with a bayonet and you can see that immediately, which tells you they can still shoot you even if they screw up on stabbing you.
i TRIED to add a wide swing visual for the bayonets except it just didn't want to work.
 
> This is more concerning than you might expect. If there exists a bayonet-attachable weapon that has a small wide-swing angle and an attack speed slower than 1 attack per second, then it becomes the case that attaching that weapon to a gun is the optimal way to wield it, because it will gain a larger wide-swing arc and faster attack speed since those are hardcoded.

the increased wide swing angle is a side effect of it being attached to a gun, but if it is a serious problem i can change it. i wasn't thinking about the attack speed, so i'll cap it at 1.

> I am not unaware of what the PR is about, my point is that if it's just about bayonets, it's not being conscious of future content. A proper attachment system may be complex, but not convoluted. In all honestly, the implementation here is more fitting of the word convoluted. This PRs issues should not be hidden behind claims of scope.

alright. i won't draft this but i'll still try and make a functional and intuitive attachments system.

**Comment by RiceMar1244** [comment]
> alright. i won't draft this but i'll still try and make a functional and intuitive attachments system.

Only if you want to. It's not on you to work on anything youre not enthusiastic about.  We're all volunteers and no one can make you take on a larger project, especially not me. I only warn of my own assessment and opinion. When the maintainers get around to reviewing this, listen to them over me.

**Comment by notquitehadouken** [comment]
> > alright. i won't draft this but i'll still try and make a functional and intuitive attachments system.
> 
> Only if you want to. It's not on you to work on anything youre not enthusiastic about. We're all volunteers and no one can make you take on a larger project, especially not me. I only warn of my own assessment and opinion. When the maintainers get around to reviewing this, listen to them over me.

i worked on it for like 5 seconds before realizing i was just making item slots but worse

**Comment by JoeHammad1844** [comment]
no visuals?

**Comment by notquitehadouken** [comment]
> no visuals?

_i tried_

**Comment by CatMagic1** [comment]
> no visuals?

he tried

womp womp 🎺

**Comment by notquitehadouken** [comment]
what even is this error

**Comment by notquitehadouken** [comment]
i love you notquitehadouken:fix_bayonet but my implementation is janky ahh and barely works

**Comment by CatMagic1** [comment]
NOOO

https://github.com/space-wizards/space-station-14/assets/135462166/ebc30625-374a-4005-ade1-3e70cd25de75

**Comment by Luxzhv** [comment]
I'd say this COULD be applied to the surplus rifle exclusively to make it be not such a shit option but... Ehhhh, I'm mostly looking at the sprite.

**Comment by JoeHammad1844** [comment]
Tragedy

**Comment by notquitehadouken** [comment]
> NOOO
> 
>  sad-cry.mp4

i'm sorry jim

**Comment by JoeHammad1844** [comment]
So have you given up or will you try again later?

**Comment by notquitehadouken** [comment]
> So have you given up or will you try again later?

i will try again because _i want my damn bayonets_
probably not soon though

**Pull request closed without merging**